### PR TITLE
Fluid typography: convert server-side block support values

### PR DESCRIPTION
**Pull request by ramonjd**
Parent issue:

- https://github.com/WordPress/gutenberg/issues/44758

## What?

Let's do this! 🕺 

For server  rendered typography block support styles, covert font size values to fluid values in the inline styles if fluid typography is activated.

❗ This has effect on the frontend only and not the editor.

## Why?

Fluid typography for **font size presets** was added in https://github.com/WordPress/gutenberg/pull/39529

However, custom font sizes should also be automatically converted to fluid values. This is a bug.

## How?
Using the existing `gutenberg_get_typography_font_size_value()` function.

## Testing Instructions

Run `npm run test:unit:php -- --filter WP_Block_Supports_Typography_Test`

Enable fluid typography via theme.json
<details>

<summary>Here is some test theme.json</summary>

```json

{
	"version": 2,
	"settings": {
		"appearanceTools": true,
		"typography": {
			"fluid": true
		}
	}
}

```

</details>

Testing in the post and site editors, add a Site Title block

<details>

<summary>Here is some example HTML</summary>

```html
<!-- wp:site-title {"style":{"typography":{"fontSize":"62px"}}} /-->

```

</details>

Apply a custom font size to the block and publish the page

Screen Shot 2022-10-07 at 12 33 05 pm

In the frontend, check that the inline font size style value has been converted to fluid

Screen Shot 2022-10-07 at 11 44 15 am

https://user-images.githubusercontent.com/6458278/194447642-01caf387-22ac-4acd-92a0-69fe09fca59f.mp4

### CODE REVIEW

File: lib/block-supports/typography.php
Comment (ramonjd) [inline lib/block-supports/typography.php]: Ah, we don't actually need this line since `gutenberg_get_typography_font_size_value()` checks explicitly for `false`
```suggestion
```

Comment (andrewserong) [reply lib/block-supports/typography.php]: Yes, good find! Let's ditch it (I think I got a bit confused when I encountered that earlier)

**Review by andrewserong — APPROVED** [review]
Nice one, this is working well for me @ramonjd, and the tests look good 👍

✅ Setting font sizes by preset still works as expected
✅ Setting a custom font size with `fluid` set to `false` results in the direct font-size value being used
✅ Switching `settings.typography.fluid` to `true` adjusts the output for server-rendered blocks to use the clampified values

| fluid `false` | fluid `true` |
| --- | --- |
| image | image |

Since this PR doesn't cover the editor behaviour / rendering, we might want to hold off on merging until that change is ready to go too? But otherwise this LGTM! ✨

**Comment by ramonjd** [comment]
> we might want to hold off on merging until that change is ready to go too?

Good point. Thanks for testing!

**Review by tellthemachines — APPROVED** [review]
I tested all the server rendered blocks with typography support and they're all working well, except for Nav Link, Submenu, Page List and Search, for which I pushed some custom logic so they should be working now!

Cover block when server rendered (if it has a featured image) doesn't seem to come under this logic. I think we'd best re-test when the work to support non-server rendered blocks is in place, as it might just work then.

**Comment by jasmussen** [comment]
Nice one! Just confirming that this fixes it on the frontend for me, even if not in the editor, which I understand is the goal for this PR:

test

File: packages/block-library/src/search/index.php
Comment (noisysocks) [inline packages/block-library/src/search/index.php]: Why does this use `esc_attr`? `$attributes['style']['typography']['fontSize']` isn't a HTML attribute.

Comment (andrewserong) [reply packages/block-library/src/search/index.php]: From memory, the search block needs to explicitly call `esc_attr` because it isn't using `get_block_wrapper_attributes`. When the other server-rendered blocks pass their styles through to that function, it winds up being escaped around this line in core: https://github.com/WordPress/wordpress-develop/blob/5ebe28966e5decbe1862c147bf98db0d8094038e/src/wp-includes/class-wp-block-supports.php#L213 — so the search block needs to do its own escaping before injecting.

Comment (noisysocks) [reply packages/block-library/src/search/index.php]: It looks like the Search block's `style` attributes are already being sanitised here using `safecss_filter_attr`:

https://github.com/WordPress/gutenberg/blob/6a2d304b3c66af697b4f535708c23e51f057115a/packages/block-library/src/search/index.php#L402-L405

That should mean all of the `esc_attr` calls in this function are unnecessary albeit probably harmless.

At the very least let's change this so that `esc_attr` is called on the output of `gutenberg_get_typography_font_size_value`, not the input. It's not right at all to escape the input as `gutenberg_get_typography_font_size_value` wouldn't know what to do with an HTML escape sequence (e.g. `$lt;` or `$quot;`) if it encountered one.

Comment (andrewserong) [reply packages/block-library/src/search/index.php]: > At the very least let's change this so that esc_attr is called on the output of gutenberg_get_typography_font_size_value, not the input.

Ah, yes, good point! Agreed.

> It looks like the Search block's style attributes are already being sanitised here using safecss_filter_attr:

`safecss_filter_attr` doesn't handle all circumstances for when a CSS string is used in an inline style, so we still need the `esc_attr` call.

Comment (noisysocks) [reply packages/block-library/src/search/index.php]: > `safecss_filter_attr` doesn't handle all circumstances for when a CSS string is used in an inline style, so we still need the `esc_attr` call.

I'd say move the `esc_attr` call up to `styles_for_block_core_search` then. Strangely that's how it works for `'label'` but none of the other keys.

Comment (tellthemachines) [reply packages/block-library/src/search/index.php]: Good feedback, hadn't thought that `esc_attr` might return something the font size function can't process 😅 I'll fix  it!

Comment (tellthemachines) [reply packages/block-library/src/search/index.php]: Done

Comment (andrewserong) [reply packages/block-library/src/search/index.php]: Thanks for fixing that up @tellthemachines! 🙇

**Comment by ndiego** [comment]
@ramonjd @noisysocks @andrewserong I am seeing a couple of approvals. Are we in good shape to merge this? If so, I want to ensure we get this backported ahead of RC1. Thanks!

**Comment by jffng** [comment]
Once more, gave this a spin with TT3 and it's working. I noticed #44764 also cover the site title test case, but this one is needed for the Nav block. Not sure technically why that is the case, but this PR seems to be necessary for some server side blocks 👍

**Comment by ramonjd** [comment]
@jffng Thanks a lot for testing these 🙇

**Comment by ockham** [comment]
Removing the "Backport to WP Beta/RC" label since this has been backported as part of #44868 (and sync'd to Core per https://github.com/WordPress/wordpress-develop/pull/3437).

**Pull request merged**
